tiledb/cloud/array.py

```diff
@@ -182,3 +186,177 @@ def array_activity(uri):
+    if not parsed.scheme == "tiledb":
+        raise Exception("Incorrect array uri, must be in tiledb:// scheme")
+    return parsed.netloc, parsed.path[1:]
+
+
+def parse_ranges(ranges):
+    """
+    Takes a list of the following objects per dimension:
+
+    - scalar index
+    - (start,end) tuple
+    - list of either of the above types
+
+    :param ranges: list of (scalar, tuple, list)
+    :param builder: function taking arguments (dim_idx, start, end)
+    :return:
+    """
+
+    def make_range(dim_range):
+        if isinstance(dim_range, (int, float)):
+            start, end = dim_range, dim_range
+        elif isinstance(dim_range, (tuple, list)):
+            start, end = dim_range[0], dim_range[1]
+        elif isinstance(dim_range, slice):
+            assert dim_range.step is None, "slice steps are not supported!"
+            start, end = dim_range.start, dim_range.stop
+        else:
+            raise ValueError("Unknown index type! (type: '{}')".format(type(dim_range)))
+        return [start, end]
+
+    result = list()
+    for dim_idx, dim_range in enumerate(ranges):
+        dim_list = []
+        # TODO handle numpy scalars here?
+        if isinstance(dim_range, (int, float, tuple, slice)):
+            dim_list.extend(make_range(dim_range))
+        elif isinstance(dim_range, list):
+            for r in dim_range:
+                dim_list.extend(make_range(r))
+        else:
+            raise ValueError(
+                "Unknown subarray/index type! (type: '{}', "
+                ", idx: '{}', value: '{}')".format(type(dim_range), dim_idx, dim_range)
+            )
+        result.append(dim_list)
+
+    return result
+
+
+def apply_async(
+    uri,
+    func,
+    ranges,
+    attrs=None,
+    layout=None,
+    image_name=None,
+    http_compressor="deflate",
+):
+    """
+    Apply a user defined function to an array asynchronous
+
+    **Example**
+    >>> import tiledb, tiledb.cloud, numpy
+    >>> def median(df):
+    ...   return numpy.median(df["a"])
+    >>> # Open the array then run the UDF
+    >>> tiledb.cloud.array.apply_async("tiledb://TileDB-Inc/quickstart_dense", median, [(0,5), (0,5)], attrs=["a", "b", "c"]).get()
+    2.0
+
+    :param func: user function to run
+    :param ranges: ranges to issue query on
+    :param attrs: list of attributes or dimensions to fetch in query
+    :param layout: tiledb query layout
+    :param image_name: udf image name to use, useful for testing beta features
+    :param http_compressor: set http compressor for results
+    :return: UDFResult object which is a future containing the results of the UDF
+    """
+
+    (namespace, array_name) = split_uri(uri)
+    api_instance = client.client.udf_api
+
+    if not callable(func):
+        raise TypeError("First argument to `apply` must be callable!")
+
+    pickledUDF = cloudpickle.dumps(func, protocol=udf.tiledb_cloud_protocol)
+    pickledUDF = base64.b64encode(pickledUDF).decode("ascii")
```

Comment:
I probably asked about this before, but it still looks weird (call `.decode('ascii')`) -- is this because the generated API doesn't accept bytes?

Reply:
Yes unfortunately it does not accept bytes, somewhere in the middle it tries to perform some operations which isn't supported on bytes. I don't have the exact details at the moment but can reproduce if you want.

Reply:
No need, understood.